src/asyncrpcqueue.cpp

```diff
@@ -0,0 +1,233 @@
+ *
+ * To retain polymorphic behaviour, i.e. main() method of derived classes is invoked,
+ * caller should create the shared_ptr like this:
+ *
+ * std::shared_ptr<AsyncRPCOperation> ptr(new MyCustomAsyncRPCOperation(params));
+ *
+ * Don't use std::make_shared<AsyncRPCOperation>().
+ */
+void AsyncRPCQueue::addOperation(const std::shared_ptr<AsyncRPCOperation> &ptrOperation) {
+    std::lock_guard<std::mutex> guard(lock_);
+
+    // Don't add if queue is closed or finishing
+    if (isClosed() || isFinishing()) {
+        return;
+    }
+
+    AsyncRPCOperationId id = ptrOperation->getId();
+    operation_map_.emplace(id, ptrOperation);
+    operation_id_queue_.push(id);
+    this->condition_.notify_one();
+}
+
+/**
+ * Return the operation for a given operation id.
+ */
+std::shared_ptr<AsyncRPCOperation> AsyncRPCQueue::getOperationForId(AsyncRPCOperationId id) const {
+    std::shared_ptr<AsyncRPCOperation> ptr;
+
+    std::lock_guard<std::mutex> guard(lock_);
+    AsyncRPCOperationMap::const_iterator iter = operation_map_.find(id);
+    if (iter != operation_map_.end()) {
+        ptr = iter->second;
+    }
+    return ptr;
+}
+
+/**
+ * Return the operation for a given operation id and then remove the operation from internal storage.
+ */
+std::shared_ptr<AsyncRPCOperation> AsyncRPCQueue::popOperationForId(AsyncRPCOperationId id) {
+    std::shared_ptr<AsyncRPCOperation> ptr = getOperationForId(id);
+    if (ptr) {
+        std::lock_guard<std::mutex> guard(lock_);
+        // Note: if the id still exists in the operationIdQueue, when it gets processed by a worker
+        // there will no operation in the map to execute, so nothing will happen.
+        operation_map_.erase(id);
+    }
+    return ptr;
+}
+
+/**
+ * Return true if the queue is closed to new operations.
+ */
+bool AsyncRPCQueue::isClosed() const {
+    return closed_.load();
+}
+
+/**
+ * Close the queue and cancel all existing operations
+ */
+void AsyncRPCQueue::close() {
+    closed_.store(true);
+    cancelAllOperations();
+}
+
+/**
+ * Return true if the queue is finishing up
+ */
+bool AsyncRPCQueue::isFinishing() const {
+    return finish_.load();
+}
+
+/**
+ * Close the queue but finish existing operations. Do not accept new operations.
+ */
+void AsyncRPCQueue::finish() {
+    finish_.store(true);
+}
+
+/**
+ *  Call cancel() on all operations
+ */
+void AsyncRPCQueue::cancelAllOperations() {
+    std::lock_guard<std::mutex> guard(lock_);
+    for (auto key : operation_map_) {
+        key.second->cancel();
+    }
+    this->condition_.notify_all();
+}
+
+/**
+ * Return the number of operations in the queue
+ */
+size_t AsyncRPCQueue::getOperationCount() const {
+    std::lock_guard<std::mutex> guard(lock_);
+    return operation_id_queue_.size();
+}
+
+/**
+ * Spawn a worker thread
+ */
+void AsyncRPCQueue::addWorker() {
+    std::lock_guard<std::mutex> guard(lock_);
+    workers_.emplace_back( std::thread(&AsyncRPCQueue::run, this, ++workerCounter) );
```

Comment:
This is the only place ``workerCounter`` is used. Also, there's no operation to remove a worker, and ``lock_`` is held here. So it could be replaced with ``workers_.size()``. (Does not block merge.)